<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />




<title>Hierarchical Generalized Additive Models</title>

<script src="site_libs/header-attrs-2.27/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/bootstrap.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.13.2/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="style.css" type="text/css" />



<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}

.tocify-subheader {
  display: inline;
}
.tocify-subheader .tocify-item {
  font-size: 0.95em;
}

</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html"></a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Data Exploration
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="01-A-data-exploration.html">Lab</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Linear Models
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="02-A-linear-models-and-probability-distributions.html">Lab</a>
    </li>
    <li>
      <a href="02-B-Homework-01.html">Homework 1</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Model Selection
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="03-A-model-selection.html">Lab</a>
    </li>
    <li>
      <a href="03-B-Homework-02.html">Homework 2</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Autocorrelation
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="04-A-Autocorrelation.html">Lab</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Mixed Models
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="05-A-Linear-Mixed-Models.html">Lab: Linear Mixed Models</a>
    </li>
    <li>
      <a href="05-B-Generalized-Linear-Mixed-Models.html">Lab: Generalized Linear Mixed Models</a>
    </li>
    <li>
      <a href="05-C-Homework-03.html">Homework 3</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    HGAMs
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="06-A-Hierarchical-Generalized-Additive-Models.html">Lab: HGAMs</a>
    </li>
    <li>
      <a href="06-B-Homework-04.html">Homework 4</a>
    </li>
  </ul>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Hierarchical Generalized Additive
Models</h1>

</div>


<p><strong>Assigned Reading:</strong></p>
<ul>
<li>Pedersen, E. J., Miller, D. L., Simpson, G. L., &amp; Ross, N.
(2019). Hierarchical generalized additive models in ecology: an
introduction with mgcv. PeerJ, 7, e6876.</li>
<li>Lawton, D., Scarth, P., Deveson, E., Piou, C., Spessa, A., Waters,
C., &amp; Cease, A. J. (2022). Seeing the locust in the swarm:
accounting for spatiotemporal hierarchy improves ecological models of
insect populations. Ecography, 2022(2).</li>
</ul>
<div id="overview" class="section level2">
<h2>Overview</h2>
<p>Invasive plants cause immense ecological and economic damage in U.S.
forested lands. The U.S. Forest Service started collecting invasive
species percent cover data</p>
<p>
<div class="float">
<img
src="https://github.com/LivingLandscapes/Course_EcologicalModeling/raw/master/images/Cogongrass_Alabama.jpg"
width="600" alt="‘Cogongrass’ by Nancy Loewenstein and John McGuire." />
<div class="figcaption">‘Cogongrass’ by Nancy Loewenstein and John
McGuire.</div>
</div>
<p>
<p>A brief description of the data:</p>
<table>
<colgroup>
<col width="21%" />
<col width="21%" />
<col width="57%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Data File Name</th>
<th align="left"><strong>Column Name</strong></th>
<th align="left"><strong>Description</strong></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FIA_Invasives_GrowthTaxonomy.csv</td>
<td align="left">SCIENTIFIC_NAME</td>
<td align="left">Genus and species</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">GrowthHabit1</td>
<td align="left">Primary growth habit</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">GrowthHabit2</td>
<td align="left">Secondary growth habit, if applicable</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">GrowthHabit3</td>
<td align="left">Tertiary growth habit, if applicable</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">GrowthHabit_Number</td>
<td align="left">Number of growth habits</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Order</td>
<td align="left">Order of the species</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">COMMON_NAME</td>
<td align="left">Species’ common name</td>
</tr>
<tr class="even">
<td align="left">FIA_Invasives_Cover.csv</td>
<td align="left">Year</td>
<td align="left">years 2001 - 2020</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">pltID</td>
<td align="left">Unique identifier for each FIA plot</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">SCIENTIFIC_NAME</td>
<td align="left">Genus and species</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">COMMON_NAME</td>
<td align="left">Species’ common name</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">cover</td>
<td align="left">areal cover of invasive species (acres)</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">Longitude</td>
<td align="left">decimal degrees longitude</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Latitude</td>
<td align="left">decimal degrees latitude</td>
</tr>
</tbody>
</table>
<hr />
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># List of packages necessary to run this script:</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">require</span>(librarian, <span class="at">quietly =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">shelf</span>(tidyverse, </span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a>      mgcv, <span class="co"># For checking model convergence</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>      MuMIn, <span class="co"># for model selection</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a>      gratia, <span class="co"># for ggplot functionality with mgcv</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a>      modelr, <span class="co"># for cross-validation</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>      vroom, <span class="co"># for loading data FAST</span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a>      <span class="at">lib =</span> <span class="fu">tempdir</span>(),</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a>      <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co"># Set the web address where R will look for files from this repository</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># Do not change this address</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>repo_url <span class="ot">&lt;-</span> <span class="st">&quot;https://raw.githubusercontent.com/LivingLandscapes/Course_EcologicalModeling/master/&quot;</span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co"># Load invasive species cover data</span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a>fia <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="fu">paste0</span>(repo_url, <span class="st">&quot;data/FIA_Invasives_Cover.csv&quot;</span>))</span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="co"># Load invasive species metadata </span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a>spp_info <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="fu">paste0</span>(repo_url, <span class="st">&quot;data/FIA_Invasives_GrowthTaxonomy.csv&quot;</span>))</span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a></span>
<span id="cb1-22"><a href="#cb1-22" tabindex="-1"></a><span class="co"># Join the cover and metadata</span></span>
<span id="cb1-23"><a href="#cb1-23" tabindex="-1"></a>fia <span class="ot">&lt;-</span> <span class="fu">left_join</span>(fia, spp_info) <span class="sc">%&gt;%</span></span>
<span id="cb1-24"><a href="#cb1-24" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">GrowthHabit1 =</span> <span class="fu">factor</span>(GrowthHabit1),</span>
<span id="cb1-25"><a href="#cb1-25" tabindex="-1"></a>         <span class="at">Order =</span> <span class="fu">factor</span>(Order),</span>
<span id="cb1-26"><a href="#cb1-26" tabindex="-1"></a>         <span class="at">cover =</span> cover <span class="sc">+</span> <span class="fl">0.0001</span>)</span>
<span id="cb1-27"><a href="#cb1-27" tabindex="-1"></a></span>
<span id="cb1-28"><a href="#cb1-28" tabindex="-1"></a><span class="co"># Take a quick look at the data</span></span>
<span id="cb1-29"><a href="#cb1-29" tabindex="-1"></a><span class="fu">head</span>(fia)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 12
##    Year pltID SCIENTIFIC_NAME COMMON_NAME  cover Longitude Latitude GrowthHabit1
##   &lt;dbl&gt; &lt;chr&gt; &lt;chr&gt;           &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;       
## 1  2001 3_1_… Albizia julibr… silktree    0.267      -86.5     32.5 Tree        
## 2  2001 3_1_… Lonicera japon… Japanese h… 0.438      -86.5     32.5 Vine        
## 3  2001 3_1_… Lonicera japon… Japanese h… 0.433      -86.5     32.4 Vine        
## 4  2001 3_1_… Melia azedarach Chinaberry… 0.0501     -86.5     32.4 Tree        
## 5  2001 3_1_… Lonicera japon… Japanese h… 0.375      -86.6     32.7 Vine        
## 6  2001 3_1_… Pueraria monta… kudzu       0.468      -86.6     32.7 Vine        
## # ℹ 4 more variables: GrowthHabit2 &lt;chr&gt;, GrowthHabit3 &lt;chr&gt;,
## #   GrowthHabit_Number &lt;dbl&gt;, Order &lt;fct&gt;</code></pre>
<div id="data-exploration" class="section level3">
<h3>Data Exploration</h3>
<p><strong>On your own, familiarize yourself with the data.</strong>
Below, you can see that we’re deadling with &gt;100k rows and
<em>many</em> potential grouping variables.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co"># Number of columns and rows?</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">dim</span>(fia) <span class="co"># &gt;100k rows!</span></span></code></pre></div>
<pre><code>## [1] 101979     12</code></pre>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># How many unique values in each column?</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>fia <span class="sc">%&gt;%</span></span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="fu">across</span>(<span class="fu">everything</span>(), <span class="sc">~</span> <span class="fu">length</span>(<span class="fu">unique</span>(.x))))</span></code></pre></div>
<pre><code>## # A tibble: 1 × 12
##    Year pltID SCIENTIFIC_NAME COMMON_NAME cover Longitude Latitude GrowthHabit1
##   &lt;int&gt; &lt;int&gt;           &lt;int&gt;       &lt;int&gt; &lt;int&gt;     &lt;int&gt;    &lt;int&gt;        &lt;int&gt;
## 1    21 29994              74          74 16069     29955    29916            6
## # ℹ 4 more variables: GrowthHabit2 &lt;int&gt;, GrowthHabit3 &lt;int&gt;,
## #   GrowthHabit_Number &lt;int&gt;, Order &lt;int&gt;</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="co"># Some preparations for mgcv</span></span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>fia <span class="ot">&lt;-</span></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>  fia <span class="sc">%&gt;%</span></span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(Order, GrowthHabit1, GrowthHabit2, GrowthHabit3, SCIENTIFIC_NAME, pltID),</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>                as.factor)) <span class="sc">%&gt;%</span> <span class="co"># mgcv wants factors for random effect</span></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">cover =</span> cover <span class="sc">+</span> <span class="fl">0.001</span>, <span class="co"># for log link</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>         ) <span class="sc">%&gt;%</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="fu">across</span>(<span class="fu">c</span>(Year, Latitude, Longitude),</span>
<span id="cb7-9"><a href="#cb7-9" tabindex="-1"></a>                <span class="sc">~</span> (.x <span class="sc">-</span> <span class="fu">mean</span>(.x)) <span class="sc">/</span> <span class="fu">sd</span>(.x),</span>
<span id="cb7-10"><a href="#cb7-10" tabindex="-1"></a>                <span class="at">.names =</span> <span class="st">&quot;{.col}_scaled&quot;</span>)) <span class="co"># scale covariates</span></span>
<span id="cb7-11"><a href="#cb7-11" tabindex="-1"></a></span>
<span id="cb7-12"><a href="#cb7-12" tabindex="-1"></a><span class="fu">head</span>(fia)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 15
##    Year pltID SCIENTIFIC_NAME COMMON_NAME  cover Longitude Latitude GrowthHabit1
##   &lt;dbl&gt; &lt;fct&gt; &lt;fct&gt;           &lt;chr&gt;        &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;fct&gt;       
## 1  2001 3_1_… Albizia julibr… silktree    0.268      -86.5     32.5 Tree        
## 2  2001 3_1_… Lonicera japon… Japanese h… 0.439      -86.5     32.5 Vine        
## 3  2001 3_1_… Lonicera japon… Japanese h… 0.434      -86.5     32.4 Vine        
## 4  2001 3_1_… Melia azedarach Chinaberry… 0.0511     -86.5     32.4 Tree        
## 5  2001 3_1_… Lonicera japon… Japanese h… 0.376      -86.6     32.7 Vine        
## 6  2001 3_1_… Pueraria monta… kudzu       0.469      -86.6     32.7 Vine        
## # ℹ 7 more variables: GrowthHabit2 &lt;fct&gt;, GrowthHabit3 &lt;fct&gt;,
## #   GrowthHabit_Number &lt;dbl&gt;, Order &lt;fct&gt;, Year_scaled &lt;dbl&gt;,
## #   Latitude_scaled &lt;dbl&gt;, Longitude_scaled &lt;dbl&gt;</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># </span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="co"># fit1 &lt;- </span></span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="co">#   bam(cover ~ te(Longitude, Latitude) + </span></span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a><span class="co">#               te(Longitude, Latitude, GrowthHabit1, bs = c(&quot;tp&quot;, &quot;tp&quot;, &quot;re&quot;)) +</span></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a><span class="co">#               s(Year),</span></span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a><span class="co">#       family = Gamma(&quot;log&quot;),</span></span>
<span id="cb9-7"><a href="#cb9-7" tabindex="-1"></a><span class="co">#       data = fia,</span></span>
<span id="cb9-8"><a href="#cb9-8" tabindex="-1"></a><span class="co">#       discrete=TRUE,</span></span>
<span id="cb9-9"><a href="#cb9-9" tabindex="-1"></a><span class="co">#       method=&quot;fREML&quot;)</span></span>
<span id="cb9-10"><a href="#cb9-10" tabindex="-1"></a><span class="co"># summary(fit1)</span></span>
<span id="cb9-11"><a href="#cb9-11" tabindex="-1"></a><span class="co"># # draw(fit1)</span></span>
<span id="cb9-12"><a href="#cb9-12" tabindex="-1"></a><span class="co"># fit2 &lt;- </span></span>
<span id="cb9-13"><a href="#cb9-13" tabindex="-1"></a><span class="co">#   bam(cover ~ te(Longitude, Latitude) + </span></span>
<span id="cb9-14"><a href="#cb9-14" tabindex="-1"></a><span class="co">#         te(Longitude, Latitude, GrowthHabit1, bs = c(&quot;tp&quot;, &quot;tp&quot;, &quot;re&quot;)),</span></span>
<span id="cb9-15"><a href="#cb9-15" tabindex="-1"></a><span class="co">#       family = Gamma(&quot;log&quot;),</span></span>
<span id="cb9-16"><a href="#cb9-16" tabindex="-1"></a><span class="co">#       data = fia,</span></span>
<span id="cb9-17"><a href="#cb9-17" tabindex="-1"></a><span class="co">#       discrete=TRUE,</span></span>
<span id="cb9-18"><a href="#cb9-18" tabindex="-1"></a><span class="co">#       method=&quot;fREML&quot;)</span></span>
<span id="cb9-19"><a href="#cb9-19" tabindex="-1"></a><span class="co"># summary(fit2)</span></span>
<span id="cb9-20"><a href="#cb9-20" tabindex="-1"></a><span class="co"># fit3 &lt;- </span></span>
<span id="cb9-21"><a href="#cb9-21" tabindex="-1"></a><span class="co">#   bam(cover ~ te(Longitude, Latitude) + </span></span>
<span id="cb9-22"><a href="#cb9-22" tabindex="-1"></a><span class="co">#         te(Longitude, Latitude, by = GrowthHabit1) +</span></span>
<span id="cb9-23"><a href="#cb9-23" tabindex="-1"></a><span class="co">#         GrowthHabit1,</span></span>
<span id="cb9-24"><a href="#cb9-24" tabindex="-1"></a><span class="co">#       family = Gamma(&quot;log&quot;),</span></span>
<span id="cb9-25"><a href="#cb9-25" tabindex="-1"></a><span class="co">#       data = fia,</span></span>
<span id="cb9-26"><a href="#cb9-26" tabindex="-1"></a><span class="co">#       discrete=TRUE,</span></span>
<span id="cb9-27"><a href="#cb9-27" tabindex="-1"></a><span class="co">#       method=&quot;fREML&quot;)</span></span>
<span id="cb9-28"><a href="#cb9-28" tabindex="-1"></a><span class="co"># summary(fit3)</span></span>
<span id="cb9-29"><a href="#cb9-29" tabindex="-1"></a><span class="co"># fit4 &lt;- </span></span>
<span id="cb9-30"><a href="#cb9-30" tabindex="-1"></a><span class="co">#   bam(cover ~ te(Longitude, Latitude, by = GrowthHabit1) +</span></span>
<span id="cb9-31"><a href="#cb9-31" tabindex="-1"></a><span class="co">#         GrowthHabit1,</span></span>
<span id="cb9-32"><a href="#cb9-32" tabindex="-1"></a><span class="co">#       family = Gamma(&quot;log&quot;),</span></span>
<span id="cb9-33"><a href="#cb9-33" tabindex="-1"></a><span class="co">#       data = fia,</span></span>
<span id="cb9-34"><a href="#cb9-34" tabindex="-1"></a><span class="co">#       discrete=TRUE,</span></span>
<span id="cb9-35"><a href="#cb9-35" tabindex="-1"></a><span class="co">#       method=&quot;fREML&quot;)</span></span>
<span id="cb9-36"><a href="#cb9-36" tabindex="-1"></a><span class="co"># summary(fit4)</span></span>
<span id="cb9-37"><a href="#cb9-37" tabindex="-1"></a><span class="co"># lapply(list(fit1, fit2, fit3, fit4), AICc)</span></span>
<span id="cb9-38"><a href="#cb9-38" tabindex="-1"></a><span class="co"># # appraise(fit2)</span></span></code></pre></div>
<!-- ### "Random Intercept": Model G -->
<!-- Let's first try the simplest HGAM: estimate a single global function for our covariate of interest (Year) plus a individual-level random effect intercept for ecoregion. **On your own, investigate each model diagnostic call (commented out in the chunk below) and interpret.** Look at the R help for more information, and ask Caleb if you have more questions. -->
<!-- ```{r eval = TRUE, message = FALSE, include = TRUE} -->
<!-- # Create model G using bam() -->
<!-- rh_modG <-  -->
<!--   bam(RH_per_min ~ s(Year) +  -->
<!--         s(Day, bs = "cc") + # Note cubic regression spline -->
<!--         te(Long, Lat) + # tensor smooth lets latitude/longitude 'interact' -->
<!--         s(Veg_Type, bs = "re"), # Random effect for ecoregion -->
<!--       family = gaussian("log"), # Why this link?! -->
<!--       data = weather, -->
<!--       method = "fREML") # need to use fREML for fast fitting. -->
<!-- # # Some basic model diagnostics -->
<!-- # summary(rh_modG) -->
<!-- # gam.check(rh_modG) -->
<!-- # gratia::appraise(rh_modG) -->
<!-- # # Use gratia::draw to show effects -->
<!-- # gratia::draw(rh_modG) -->
<!-- ``` -->
<!-- Okay, so the model isn't the greatest for a few reasons (can you list some?). However, for fun--and so we can see how "random intercepts" work with HGAMs--let's plot some predictions. We'll also do a quick comparison of fitted vs. observed to gauge how well the models are performing. -->
<!-- ```{r eval = TRUE, message = FALSE, include = TRUE} -->
<!-- # setup prediction data. This is a little tricky because for ggplot, we only -->
<!-- # want one latitude/longitude for each ecoregion AND we want to make sure the -->
<!-- # correct latitude/longitudes get paired with their respective ecoregions. Doing -->
<!-- # this with a left_join() -->
<!-- rh_pred_df <- -->
<!--   with( -->
<!--     weather, -->
<!--     left_join( -->
<!--       expand.grid( -->
<!--         Year = 2010:2019, -->
<!--         Day = median(Day),  # Only using the median day here! -->
<!--         Veg_Type = unique(Veg_Type) -->
<!--       ), -->
<!--       weather %>% -->
<!--         filter(Day == median(Day)) %>% -->
<!--         group_by(Veg_Type) %>% -->
<!--         filter(Station_ID %in% sample(Station_ID, 1, replace = FALSE)) %>% -->
<!--         select(c(Veg_Type, Lat, Long, Year, RH_per_min)), -->
<!--      keep = FALSE -->
<!--     ) -->
<!--   ) -->
<!-- # make the prediction, add this and a column of standard errors to the prediction -->
<!-- # data.frame. Predictions are on the log scale. -->
<!-- rh_modG_pred <- cbind(rh_pred_df, -->
<!--                        predict(rh_modG,  -->
<!--                                rh_pred_df,  -->
<!--                                se.fit = TRUE,  -->
<!--                                type = "response")) -->
<!-- # make the prediction plots -->
<!-- ggplot(data = rh_modG_pred) + -->
<!--   facet_wrap(~ Veg_Type) + -->
<!--   geom_ribbon(aes(ymin = fit - 1.96 * se.fit,  -->
<!--                   ymax = fit + 1.96 * se.fit,  -->
<!--                   x = Year,  -->
<!--                   group = Veg_Type), -->
<!--               fill = "grey80") + -->
<!--   geom_line(aes(x = Year,  -->
<!--                 y = fit,  -->
<!--                 group = Veg_Type), -->
<!--             color = "darkred") + -->
<!--   scale_x_continuous(breaks = c(2010, 2013, 2016, 2019)) +  -->
<!--   labs(x = "Year", -->
<!--        y = "Minimum Relative Humidity (%)") +  -->
<!--   ggtitle("Model G") -->
<!-- # Now, let's create a data.frame comparing observed vs. fitted values -->
<!-- rh_modG_predVSobs <-  -->
<!--   transform(rh_modG_pred,  -->
<!--             modG = predict(rh_modG,  -->
<!--                            rh_modG_pred, -->
<!--                            type = "response")) -->
<!-- # Plot observed vs. fitted values. This is a "quick" way to see how well -->
<!-- # the model predicted the data. -->
<!-- ggplot(rh_modG_predVSobs, -->
<!--        aes(x = modG,  -->
<!--            y = RH_per_min)) + -->
<!--   facet_wrap(~ Veg_Type) + -->
<!--   geom_point(alpha = 0.6) + -->
<!--   geom_abline(color = "gold", -->
<!--               alpha = 0.5) + -->
<!--   labs(x="Predicted RH", y="Observed RH") +  -->
<!--   ggtitle("Model G") -->
<!-- ``` -->
<!-- **On your own, interpret these prediction plots.** Remember the values we fed the predict() function when you're interpreting. For instance, what does it mean that we only used the median Julian day in these predictions? Does it matter that we only used one randomly-selected weather station? -->
<!-- ### "Random Slope": Model GS -->
<!-- Let's try for a  -->
<!-- ```{r eval = TRUE, message = FALSE, include = TRUE} -->
<!-- # Create model GS using bam() -->
<!-- rh_modGS <- -->
<!--   bam( -->
<!--     RH_per_min ~ s(Year) + -->
<!--       s(Year, Veg_Type, bs = "fs", m = 2) + -->
<!--       s(Day, bs = "cc") + -->
<!--       te(Long, Lat), -->
<!--     family = gaussian("log"), -->
<!--     data = weather, -->
<!--     method = "fREML" -->
<!--   ) -->
<!-- # # Some basic model diagnostics -->
<!-- # summary(rh_modGS) -->
<!-- # gam.check(rh_modGS) -->
<!-- # gratia::appraise(rh_modGS) -->
<!-- # # Use gratia::draw to show effects. This is slow, so commenting out -->
<!-- # gratia::draw(rh_modGS) -->
<!-- ``` -->
<!-- **On your own, run the model diagnostics and compare to diagnostics from Model G.** Are there improvements, and if so, where do you see them? Should we consider increasing "k" for any of the smoothers? -->
<!-- To be thorough, let's also plot predictions and compare fitted to observed: -->
<!-- ```{r eval = TRUE, message = FALSE, include = TRUE} -->
<!-- # make the prediction, add this and a column of standard errors to the prediction -->
<!-- # data.frame. Predictions are on the log scale. -->
<!-- rh_modGS_pred <- cbind(rh_pred_df, -->
<!--                        predict(rh_modGS,  -->
<!--                                rh_pred_df,  -->
<!--                                se.fit = TRUE,  -->
<!--                                type = "response")) -->
<!-- # make the prediction plots -->
<!-- ggplot(data = rh_modGS_pred) + -->
<!--   facet_wrap(~ Veg_Type) + -->
<!--   geom_ribbon(aes(ymin = fit - 1.96 * se.fit,  -->
<!--                   ymax = fit + 1.96 * se.fit,  -->
<!--                   x = Year,  -->
<!--                   group = Veg_Type), -->
<!--               fill = "grey80") + -->
<!--   geom_line(aes(x = Year,  -->
<!--                 y = fit,  -->
<!--                 group = Veg_Type), -->
<!--             color = "darkred") + -->
<!--   scale_x_continuous(breaks = c(2010, 2013, 2016, 2019)) +  -->
<!--   labs(x = "Year", -->
<!--        y = "Minimum Relative Humidity (%)") +  -->
<!--   ggtitle("Model GS") -->
<!-- # Now, let's create a data.frame comparing observed vs. fitted values -->
<!-- rh_modGS_predVSobs <-  -->
<!--   transform(rh_modGS_pred,  -->
<!--             modGS = predict(rh_modGS,  -->
<!--                            rh_modGS_pred, -->
<!--                            type = "response")) -->
<!-- # Plot observed vs. fitted values. This is a "quick" way to see how well -->
<!-- # the model predicted the data. -->
<!-- ggplot(rh_modGS_predVSobs, -->
<!--        aes(x = modGS,  -->
<!--            y = RH_per_min)) + -->
<!--   facet_wrap(~ Veg_Type) + -->
<!--   geom_point(alpha = 0.6) + -->
<!--   geom_abline(color = "gold", -->
<!--               alpha = 0.5) + -->
<!--   labs(x="Predicted RH", y="Observed RH") +  -->
<!--   ggtitle("Model GS") -->
<!-- ``` -->
<!-- Hmmm. The predictions are fairly different now. Also, it looks like some predictions are tightening up (e.g., compare Flint Hills fitted vs. observed between Models G and GS) while others got worse (can you spot any?). -->
<!-- ### Comparing HGAMs -->
<!-- We learned from the Pedersen et al. (2019) paper that we can use AIC to compare HGAMs, which is awesome! But--model selection via AIC should not be our stopping point. In this section, we're going to dip our toes into cross validation to compare the *predictive power* of HGAMs.  -->
<!-- But first, let's just see what AIC(c) has to say about our two HGAMs: -->
<!-- ```{r eval = TRUE, message = FALSE, include = TRUE} -->
<!-- model.sel(list(GS = rh_modGS, G = rh_modG)) -->
<!-- ``` -->
<!-- We have a *very* clear winner per AICc rankings: model GS.  -->
<!-- Let's check on the cross validation.  -->
<!-- ``` {r eval = TRUE, message = FALSE, include = TRUE} -->
<!-- # Create training and testing data.frames -->
<!-- train_df <-  -->
<!--   sample_n(weather %>% mutate(id = 1:n()),  -->
<!--            ceiling(nrow(weather)/10)) -->
<!-- test_df <-  -->
<!--   anti_join(weather %>% mutate(id = 1:n()),  -->
<!--             train_df, by = "id") -->
<!-- # train models -->
<!-- rh_modG_train <-  -->
<!--   bam( -->
<!--     RH_per_min ~ s(Year) + -->
<!--       s(Day, bs = "cc") + -->
<!--       te(Long, Lat) + -->
<!--       s(Veg_Type, bs = "re"), -->
<!--     family = gaussian("log"), -->
<!--     data = train_df, -->
<!--     method = "fREML" -->
<!--   ) -->
<!-- rh_modGS_train <- -->
<!--   bam( -->
<!--     RH_per_min ~ s(Year) + -->
<!--       s(Year, Veg_Type, bs = "fs", m = 2) + -->
<!--       s(Day, bs = "cc") + -->
<!--       te(Long, Lat), -->
<!--     family = gaussian("log"), -->
<!--     data = weather, -->
<!--     method = "fREML" -->
<!--   )   -->
<!-- # Root mean square error and mean absolute error -->
<!-- data.frame(Model = c("G", "GS"), -->
<!--            RMSE = c(rmse(rh_modG_train, test_df),  -->
<!--                     rmse(rh_modGS_train, test_df)), -->
<!--            MAE = c(mae(rh_modG_train, test_df),  -->
<!--                     mae(rh_modGS_train, test_df))) -->
<!-- ``` -->
<!-- Oof! Looks like both our (quick and dirty) cross-validation tests show the models have negligible differences in predictive ability. Just goes to show that we need to assess models from multiple angles.  -->
<!-- ### Try other models! -->
<!-- **On your own, try some different model formulations (e.g., model I, model GI, etc.) from Pedersen et al. (2019).** Do you see any clear patterns? -->
</div>
</div>



</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = false;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
